### PR TITLE
implement mul & div for i128

### LINKED ISSUE #35: Implement 128-bit types necessary for NEAR blockchain runtime
Currently different types in the library are in various stages of implementation completeness.

NEAR uses 128-bit numbers internally, therefore we need following types fully implemented:
- [ ] `u128`
- [ ] `i128`
- [ ] `safe.u128`
- [ ] `safe.i128`

Acceptance criteria:
- [ ] All types listed above implement all operations currently supported by `u128`
- [ ] `safe.` types also implement overflow check for all operations which can result in overflow
- [ ] Code has 100% branch coverage in tests

### PR DESCRIPTION
**Pull request by plouis01**
Partially fix https://github.com/MaxGraey/as-bignum/issues/35

### CODE REVIEW

**Comment by plouis01** [comment]
I thought the test failures were due to my pr but they were already present last pr (cf https://github.com/MaxGraey/as-bignum/pull/81)

**Comment by MaxGraey** [comment]
Please sync with main branch and update deps. Latest commit should fix test runner

File: assembly/__tests__/i128.spec.ts [hunk 1]
```diff
@@ -90,6 +91,85 @@ describe("Basic Operations", () => {
...
+
+  it("Should invariant left shift zero number", () => {
+    var a = new i128(1, 1);
+    expect(a << 0).toStrictEqual(a);
+  });
+
+  it("Should right shift one number", () => {
+    var a = new i128(0, 100);
+    expect(a >> 65).toStrictEqual(new i128(50));
+  });
```
Comment (MaxGraey) [inline assembly/__tests__/i128.spec.ts]: I'm not sure `>>` op for `i128` is correct. Please add tests for negative values, for value without hi or lo parts and shift lenght: `16`, `32`, `63`, `64`, `127`

File: assembly/integer/i128.ts [hunk 2]
```diff
@@ -279,6 +293,60 @@ export class i128 {
...
+    var hi = ah * bh;
+    var carry = u64((lo >>> 32) + (mid1 & 0xffffffff) + (mid2 & 0xffffffff));
+    var lo64 = u64(lo) + (u64(mid1) << 32) + (u64(mid2) << 32);
+    var hi64 = hi + (carry >>> 32) + ((mid1 >>> 32) + (mid2 >>> 32));
+    return new i128((hi64 << 64) | lo64);
+  }
+
+  @inline @operator('/')
+  static div(a: i128, b: i128): i128 {
+    if (b == i128.Zero) {
```
Comment (MaxGraey) [inline assembly/integer/i128.ts]:
```suggestion
    if (b.isZero()) {
```

File: assembly/integer/i128.ts [hunk 3]
```diff
@@ -279,6 +293,60 @@ export class i128 {
...
+      // Divide by zero
+      throw new Error("Division by zero");
+    }
+    if (a == i128.Min && b == i128.One.neg()) {
```
Comment (MaxGraey) [inline assembly/integer/i128.ts]: Please add
```ts
@inline static get NegOne(): i128 { return new i128(-1, -1); }
```
field to `i128` and compare with it as
```ts
b == i128.NegOne
```

File: assembly/integer/i128.ts [hunk 4]
```diff
@@ -279,6 +293,60 @@ export class i128 {
...
+    var q = i128.Zero;
+    var r = i128.Zero;
+    for (var i = 127; i >= 0; i--) {
+      r = i128.shl(r, 1);
+      r = i128.or(r, i128.and(i128.shr(a,i), i128.One));
+      if (r >= b) {
+        r = i128.sub(r, b);
+        q = i128.or(q, i128.shl(i128.One, i));
+      }
+    }
```
Comment (MaxGraey) [inline assembly/integer/i128.ts]: You can reuse `__udivmod128` here instead inefficient per-bit division

File: assembly/__tests__/i128.spec.ts (hunk 1 above)
Comment (plouis01) [reply assembly/__tests__/i128.spec.ts]: you're right, not all tests pass for neg value, I will try to fix this when I have some time

File: assembly/integer/i128.ts [hunk 5]
```diff
@@ -379,6 +447,16 @@ export class i128 {
...
+   @inline
+   i128ToU64(): u64 {
+    return u128.fromI128(this).toU64();
+  }
+
+  @inline
+  i128toi64(): i64 {
+    return this.lo; // return the lower 64 bits of x
+  }
+
```
Comment (MaxGraey) [inline assembly/integer/i128.ts]: For such operations, we have generic `as` method. See it on `u128`:
https://github.com/MaxGraey/as-bignum/blob/master/assembly/integer/u128.ts#L840

File: assembly/integer/i128.ts [hunk 6]
```diff
@@ -279,6 +294,60 @@ export class i128 {
...
+    var bl = b.i128ToU64();
+    var lo = al * bl;
+    var mid1 = ah * bl;
+    var mid2 = al * bh;
+    var hi = ah * bh;
+    var carry = u64((lo >>> 32) + (mid1 & 0xffffffff) + (mid2 & 0xffffffff));
+    var lo64 = u64(lo) + (u64(mid1) << 32) + (u64(mid2) << 32);
+    var hi64 = hi + (carry >>> 32) + ((mid1 >>> 32) + (mid2 >>> 32));
+    return new i128((hi64 << 64) | lo64);
+  }
```
Comment (MaxGraey) [inline assembly/integer/i128.ts]: Just reuse `__multi3` here from `globals.ts`. It should be similar to `mul` in u128:
https://github.com/MaxGraey/as-bignum/blob/master/assembly/integer/u128.ts#L401

File: assembly/integer/i128.ts [hunk 7]
```diff
@@ -379,6 +448,16 @@ export class i128 {
...
    }
 
+   @inline
+   i128ToU64(): u64 {
+    return u128.fromI128(this).toU64();
+  }
+
+  @inline
+  i128toi64(): i64 {
+    return this.lo; // return the lower 64 bits of x
```
Comment (MaxGraey) [inline assembly/integer/i128.ts]: Also it's not correct due to sign bit should be inherit from hi part

File: assembly/integer/i128.ts [hunk 8]
```diff
@@ -243,16 +245,17 @@ export class i128 {
...
-      return new i128(lo, hi);
-    } else {
-      var lo = value.hi >> (shift - 64);
-      return new i128(lo, 0);
+    let sign = value.isNeg();
+
+    let res = u128.shr(u128.fromI128(i128.abs(value)),shift).toI128();
+    
+    if (sign) {  // Put back the sign
+      res = res.neg();
```
Comment (MaxGraey) [inline assembly/integer/i128.ts]: This approach some overhead. For `i128.mul` / `i128.div` this is acceptable, but for such low-level operation like `shr` it is better to operate with the native parts `lo` and `hi` without GC which implicitly produced by `i128.abs`, `toI128` and `res.neg`

File: assembly/integer/i128.ts [hunk 9]
```diff
@@ -296,18 +299,23 @@ export class i128 {
...
+    }
+    if (b < i128.Zero) {
+      b = b.neg();
+      neg = !neg;
+    }
+    var prod = new i128(__multi3(a.lo, a.hi, b.lo, b.hi),__res128_hi);
+    if (neg) {
+      prod = prod.neg();
+    }
+    return prod;
```
Comment (MaxGraey) [inline assembly/integer/i128.ts]: You don't need handle sign for multiply. Just directly use `i128(__multi3(a.lo, a.hi, b.lo, b.hi),__res128_hi)`. Multiplication is some-kind of sign-agnostic operation, unlike of division.

File: assembly/integer/i128.ts [hunk 10]
```diff
@@ -279,6 +297,56 @@ export class i128 {
...
+    }
+    if (a == i128.Min && b == i128.NegOne) {
+      // Overflow
+      throw new Error("Integer overflow");
+    }
+    if (b == i128.One) {
+      return a;
+    }
+    var neg = false;
+    if (a < i128.Zero) {
```
Comment (MaxGraey) [inline assembly/integer/i128.ts]:
```suggestion
    if (a.isNeg()) {
```

File: assembly/integer/i128.ts [hunk 11]
```diff
@@ -279,6 +297,56 @@ export class i128 {
...
+      a = a.neg();
+      neg = !neg;
+    }
+    if (b < i128.Zero) {
```
Comment (MaxGraey) [inline assembly/integer/i128.ts]:
```suggestion
    if (b.isNeg()) {
```

File: assembly/integer/i128.ts [hunk 12]
```diff
@@ -279,6 +297,56 @@ export class i128 {
...
+      b = b.neg();
+      neg = !neg;
+    }
+    var q = new i128(__udivmod128(a.lo, a.hi, b.lo, b.hi), __divmod_quot_hi);
+    if (neg) {
+      q = q.neg();
+    }
+    return q;
```
Comment (MaxGraey) [inline assembly/integer/i128.ts]:
```suggestion
    return neg ? q.neg() : q;
```

File: assembly/integer/i128.ts (hunk 8 above)
Comment (plouis01) [reply assembly/integer/i128.ts]: I didn't manage to make it work another way, I might try again later this week but I don't know if I have the time

File: assembly/integer/i128.ts [hunk 13]
```diff
@@ -279,6 +294,43 @@ export class i128 {
...
+      // Divide by zero
+      throw new Error("Division by zero");
+    }
+    if (a == i128.Min && b == i128.NegOne) {
+      // Overflow
+      throw new Error("Integer overflow");
+    }
+    if (b == i128.One) {
+      return a;
+    }
```
Comment (MaxGraey) [inline assembly/integer/i128.ts]: It's unnecessary and already handled by `__udivmod128`

File: assembly/integer/i128.ts [hunk 14]
```diff
@@ -279,6 +294,43 @@ export class i128 {
...
+      __res128_hi
+    );
+  }
+
+  @inline @operator('/')
+  static div(a: i128, b: i128): i128 {
+    if (b.isZero()) {
+      // Divide by zero
+      throw new Error("Division by zero");
+    }
```
Comment (MaxGraey) [inline assembly/integer/i128.ts]: This also can be removed due to handled by `__udivmod128`

File: assembly/__tests__/i128.spec.ts [hunk 15]
```diff
@@ -90,6 +91,114 @@ describe("Basic Operations", () => {
...
+    expect(i128.mul(a.toI128(),b.toI128())).toStrictEqual(u128.mul(a,b).toI128());
+  });
+  it("Should multiplication 2", () => {
+    let a = i128.One;
+    expect(i128.mul(a,a)).toStrictEqual(a);
+  });
+  it("Should multiplication 3", () => {
+    let a = i128.One;
+    let b = i128.add(a,a);
+    expect(i128.mul(a,b)).toStrictEqual(b);
```
Comment (MaxGraey) [inline assembly/__tests__/i128.spec.ts]:
```suggestion
    let a = i128.One;
    let b = i128.fromI32(2);
    expect(i128.mul(a,b)).toStrictEqual(b);
```

File: assembly/__tests__/i128.spec.ts [hunk 16]
```diff
@@ -90,6 +91,114 @@ describe("Basic Operations", () => {
...
+    let c = i128.add(a,b);
+    let d = i128.add(c,c);
+    expect(i128.div(d,c)).toStrictEqual(b);
+  });
+  it("Should Division 5", () => {
+    let a = i128.One;
+    let b = i128.add(a,a);
+    let c = i128.add(a,b);
+    let d = i128.add(c,c);
+    expect(i128.div(d,b)).toStrictEqual(c);
```
Comment (MaxGraey) [inline assembly/__tests__/i128.spec.ts]: Also please avoid extra `i128.add` ops and just use immediate values

File: assembly/__tests__/i128.spec.ts [hunk 17]
```diff
@@ -90,6 +91,111 @@ describe("Basic Operations", () => {
...
+    let b = i128.fromI32(2);
+    let c = i128.fromI32(3);
+    let d = i128.fromI32(6);
+    expect(i128.div(d,c)).toStrictEqual(b);
+  });
+  it("Should Division 5", () => {
+    let b = i128.fromI32(2);
+    let c = i128.fromI32(3);
+    let d = i128.fromI32(6);
+    expect(i128.div(d,b)).toStrictEqual(c);
```
Comment (MaxGraey) [inline assembly/__tests__/i128.spec.ts]: these two tests are almost identical and don't really cover anything. It would be nice to have these cases: -1 / 3, -5 / 3, -7 / -2, 7 / -3 and `min / -1` (which should throw error)

File: assembly/integer/i128.ts (hunk 8 above)
Comment (plouis01) [reply assembly/integer/i128.ts]: I don't have the time to get back on this, I might as well delete shr (and tests) to merge the rest no?

Comment (MaxGraey) [reply assembly/integer/i128.ts]: That's a good decision

**Comment by MaxGraey** [comment]
Thanks!
